servicemonitor crud接口

yaoxinwang-git opened the pull request:
1.bcs-monitor 添加servicemonitor crud接口
2.servicemonitor结构体来源于github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring v0.46.0 高版本对gomod文件修改较多

https://github.com/TencentBlueKing/bk-bcs/issues/2032

tencentblueking-adm commented:
[![CLA assistant check](https://bk-cla.bktencent.com:443/pull/badge/signed)](https://bk-cla.bktencent.com:443/TencentBlueKing/bk-bcs?pullRequest=2062) <br/>All committers have signed the CLA.

adevjoe commented on bcs-services/bcs-monitor/pkg/api/routes.go:
```diff
@@ -190,6 +191,19 @@ func registerMetricsRoutes(engine *gin.RouterGroup) {
 		route.GET("/namespaces/:namespace/pods/:pod/containers/:container/disk_write_total",
 			rest.RestHandlerFunc(metrics.ContainerDiskWriteTotal))
+
+		route.GET("/service_monitors/namespaces/:namespace",
```
`/namespaces/:namespace` 要放在前面，遵循 restful 风格

adevjoe commented on bcs-services/bcs-monitor/pkg/api/routes.go:
```diff
@@ -190,6 +191,19 @@ func registerMetricsRoutes(engine *gin.RouterGroup) {
+			rest.RestHandlerFunc(service_monitor.CreateServiceMonitor))
+		route.GET("/service_monitors/namespaces/:namespace/servicemonitors/:name",
+			rest.RestHandlerFunc(service_monitor.GetServiceMonitor))
+		route.POST("/service_monitors/update/namespaces/:namespace/servicemonitors/:name",
```
`name` 放在 body 里面，路径可以是 `/namespaces/:namespace/service_monitors`，路径里面不需要动词

adevjoe commented on bcs-services/bcs-monitor/pkg/api/routes.go:
```diff
@@ -190,6 +191,19 @@ func registerMetricsRoutes(engine *gin.RouterGroup) {
+
+		route.GET("/service_monitors/namespaces/:namespace",
+			rest.RestHandlerFunc(service_monitor.ListServiceMonitors))
+		route.POST("/service_monitors/create/namespaces/:namespace/servicemonitors/:name",
```
同更新接口

adevjoe commented on bcs-services/bcs-monitor/pkg/api/routes.go:
```diff
@@ -190,6 +191,19 @@ func registerMetricsRoutes(engine *gin.RouterGroup) {
+			rest.RestHandlerFunc(service_monitor.UpdateServiceMonitor))
+		route.DELETE("/service_monitors/namespaces/:namespace/servicemonitors/:name",
+			rest.RestHandlerFunc(service_monitor.DeleteServiceMonitor))
+		route.DELETE("/service_monitors/batchdelete",
```
batchdelete 因为有 body，不要用 DELETE，根据规范这里用 POST

adevjoe commented on bcs-services/bcs-monitor/pkg/api/servicemonitor/servicemonitor.go:
```diff
@@ -0,0 +1,359 @@
+             如 `BCS-K8S-40000:default` 会被处理成 `40000:5f03d33dde`。其中，保留集群数字 ID 的目的是用于
+             NamespaceProvider 中的 fetch_instance_info 方法
+*/
+func calcIamNsId(clusterId string, namespace string) (string, error) {
```
https://github.com/TencentBlueKing/bk-bcs/blob/master/bcs-services/pkg/bcs-auth/utils/utils.go#L173

可以用这个函数

adevjoe commented on bcs-services/bcs-monitor/pkg/api/servicemonitor/servicemonitor.go:
```diff
@@ -0,0 +1,359 @@
+	limit := c.Query("limit")
+	offset := c.Query("offset")
+
+	clusterMap, err := service_monitor.ListClusters(projectId, limit, offset)
```
不需要校验集群

adevjoe commented on bcs-services/bcs-monitor/pkg/api/servicemonitor/servicemonitor.go:
```diff
@@ -0,0 +1,359 @@
+	}
+
+	serviceMonitors := service_monitor.ListServiceMonitor("")
+	handleServiceMonitorItems(clusterId, clusterMap, namespaceMap, serviceMonitors)
```
直接返回 serviceMonitor 列表就可以了，不需要特殊处理，和详情接口一样

adevjoe commented on bcs-services/bcs-monitor/pkg/component/service_monitor/servicemonitor.go:
```diff
@@ -0,0 +1,291 @@
+}
+
+// CreateServiceMonitor 创建servicemonitor
+func CreateServiceMonitor(namespace string, spec *ServiceMonitor) (*ServiceMonitor, error) {
```
servicemonitor 的接口不用单独封装，直接用 [prometheus-operator](https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/client/versioned/clientset.go) 提供的 client 就可以

adevjoe commented on bcs-services/bcs-monitor/pkg/api/servicemonitor/servicemonitor.go:
```diff
@@ -0,0 +1,322 @@
+
+func GetMonitoringV1Client(c *rest.Context) (monitoringv1.MonitoringV1Interface, error) {
+	clusterId := c.Param("clusterId")
+	kubeConf, err := clientutil.GetKubeConfByClusterId(clusterId)
```
配置文件不同，这是不能直接调用 `bcs-unified-apiserver` 的 `GetKubeConfByClusterId ` 方法

adevjoe commented on bcs-services/bcs-monitor/pkg/api/routes.go:
```diff
@@ -190,6 +191,19 @@ func registerMetricsRoutes(engine *gin.RouterGroup) {
+			rest.RestHandlerFunc(service_monitor.GetServiceMonitor))
+		route.POST("/namespaces/:namespace/service_monitors",
+			rest.RestHandlerFunc(service_monitor.CreateServiceMonitor))
+		route.PUT("/namespaces/:namespace/service_monitors",
```
PUT 带上 `:name`

adevjoe commented on bcs-services/bcs-monitor/pkg/api/servicemonitor/servicemonitor.go:
```diff
@@ -0,0 +1,322 @@
+
+	serviceMonitor := &v1.ServiceMonitor{}
+
+	serviceMonitor.APIVersion = "monitoring.coreos.com/v1"
```
APIVersion 和 kind 不用设置，默认就有了

adevjoe commented on bcs-services/bcs-monitor/pkg/api/servicemonitor/servicemonitor.go:
```diff
@@ -0,0 +1,322 @@
+// @Success 200 {string} string
+// @Router  /service_monitors/update/namespaces/:namespace/servicemonitors/:name [post]
+func UpdateServiceMonitor(c *rest.Context) (interface{}, error) {
+	//更新会合并 selector，因此先删除, 再创建
```
为什么要删除，直接覆盖更新应该就可以吧

yaoxinwang-git committed:
1.bcs-monitor 添加servicemonitor crud接口
2.servicemonitor结构体来源于github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring v0.46.0 高版本对gomod文件修改较多
